File: apps/backend/app/schemas/neo.py
from datetime import date
from enum import Enum

from pydantic import BaseModel, Field


class NeoSortBy(str, Enum):
    """Allowed sorting fields for the NEO list endpoint."""

    DATE = "date"
    DISTANCE = "distance"
    VELOCITY = "velocity"
    DIAMETER = "diameter"


class SortOrder(str, Enum):
    """Allowed sorting directions."""

    ASC = "asc"
    DESC = "desc"


class NeoSummary(BaseModel):
    """Normalized Near Earth Object summary returned to the frontend."""

    id: str
    name: str

    nasa_jpl_url: str | None = None
    absolute_magnitude_h: float | None = None

    is_potentially_hazardous: bool

    close_approach_date: date

    miss_distance_km: float | None = Field(
        default=None,
        description="Minimum miss distance in kilometers.",
    )
    relative_velocity_kmh: float | None = Field(
        default=None,
        description="Relative velocity in kilometers per hour.",
    )

    diameter_min_m: float | None = None
    diameter_max_m: float | None = None
    diameter_avg_m: float | None = None


class NeoFeedResponse(BaseModel):
    """Response model for the NEO feed endpoint."""

    start_date: date
    end_date: date
    count: int
    results: list[NeoSummary]


class NeoCloseApproach(BaseModel):
    """Normalized close approach information for a NEO detail page."""

    close_approach_date: date | None = None
    close_approach_date_full: str | None = None
    epoch_date_close_approach: int | None = None

    relative_velocity_kmh: float | None = None
    miss_distance_km: float | None = None

    orbiting_body: str | None = None


class NeoOrbitalData(BaseModel):
    """Normalized orbital information returned by NASA NeoWs."""

    orbit_id: str | None = None
    orbit_determination_date: str | None = None

    first_observation_date: date | None = None
    last_observation_date: date | None = None

    data_arc_in_days: int | None = None
    observations_used: int | None = None

    orbit_class_type: str | None = None
    orbit_class_description: str | None = None
    orbit_class_range: str | None = None


class NeoDetailResponse(BaseModel):
    """Detailed normalized information for a single Near Earth Object."""

    id: str
    name: str
    designation: str | None = None

    nasa_jpl_url: str | None = None
    absolute_magnitude_h: float | None = None

    is_potentially_hazardous: bool
    is_sentry_object: bool

    diameter_min_m: float | None = None
    diameter_max_m: float | None = None
    diameter_avg_m: float | None = None

    orbital_data: NeoOrbitalData | None = None
    close_approaches: list[NeoCloseApproach]


class NeoDailyCount(BaseModel):
    """Number of NEOs grouped by close approach date."""

    date: date
    count: int


class NeoHazardousDistributionItem(BaseModel):
    """Chart-friendly item representing hazardous/non-hazardous distribution."""

    label: str
    value: int


class NeoStatsResponse(BaseModel):
    """Aggregated statistics for the NEO dashboard."""

    start_date: date
    end_date: date

    total_count: int
    hazardous_count: int
    non_hazardous_count: int

    average_diameter_m: float | None = None
    min_miss_distance_km: float | None = None
    max_relative_velocity_kmh: float | None = None

    closest_neo: NeoSummary | None = None
    fastest_neo: NeoSummary | None = None
    largest_neo: NeoSummary | None = None

    daily_counts: list[NeoDailyCount]
    hazardous_distribution: list[NeoHazardousDistributionItem]

File: apps/web/src/types/neo.ts
export type NeoSortBy = "date" | "distance" | "velocity" | "diameter";

export type SortOrder = "asc" | "desc";

export interface NeoSummary {
    id: string;
    name: string;

    nasa_jpl_url: string | null;
    absolute_magnitude_h: number | null;

    is_potentially_hazardous: boolean;

    close_approach_date: string;

    miss_distance_km: number | null;
    relative_velocity_kmh: number | null;

    diameter_min_m: number | null;
    diameter_max_m: number | null;
    diameter_avg_m: number | null;
}

export interface NeoFeedResponse {
    start_date: string;
    end_date: string;
    count: number;
    results: NeoSummary[];
}

export interface NeoDailyCount {
    date: string;
    count: number;
}

export interface NeoHazardousDistributionItem {
    label: string;
    value: number;
}

export interface NeoStatsResponse {
    start_date: string;
    end_date: string;

    total_count: number;
    hazardous_count: number;
    non_hazardous_count: number;

    average_diameter_m: number | null;
    min_miss_distance_km: number | null;
    max_relative_velocity_kmh: number | null;

    closest_neo: NeoSummary | null;
    fastest_neo: NeoSummary | null;
    largest_neo: NeoSummary | null;

    daily_counts: NeoDailyCount[];
    hazardous_distribution: NeoHazardousDistributionItem[];
}

export interface NeoOrbitalData {
    orbit_id: string | null;
    orbit_determination_date: string | null;

    first_observation_date: string | null;
    last_observation_date: string | null;

    data_arc_in_days: number | null;
    observations_used: number | null;

    orbit_class_type: string | null;
    orbit_class_description: string | null;
    orbit_class_range: string | null;
}

export interface NeoCloseApproach {
    close_approach_date: string | null;
    close_approach_date_full: string | null;
    epoch_date_close_approach: number | null;

    relative_velocity_kmh: number | null;
    miss_distance_km: number | null;

    orbiting_body: string | null;
}

export interface NeoDetailsResponse {
    id: string;
    name: string;
    designation: string | null;

    nasa_jpl_url: string | null;
    absolute_magnitude_h: number | null;

    is_potentially_hazardous: boolean;
    is_sentry_object: boolean;

    diameter_min_m: number | null;
    diameter_max_m: number | null;
    diameter_avg_m: number | null;

    orbital_data: NeoOrbitalData | null;
    close_approaches: NeoCloseApproach[];
}
export interface NeoFeedQuery {
    startDate: string;
    endDate: string;
    hazardous?: boolean | null;
    sortBy?: NeoSortBy;
    sortOrder?: SortOrder;
}

export interface NeoStatsQuery {
    startDate: string;
    endDate: string;
}